File: 2015/swift/6/main-part-1.swift
import Foundation

struct Action {
    var act: String
    var start: (Int, Int)
    var end: (Int, Int)
}

func get_action(_ line: String) -> Action {
    let parts = line.components(separatedBy: " ")
    if parts[0] == "turn" {
        let action = parts[1]
        let start = parts[2].components(separatedBy: ",").map { Int($0)! }
        let end = parts[4].components(separatedBy: ",").map { Int($0)! }
        return Action(act: action, start: (start[0], start[1]), end: (end[0], end[1]))
    } else {
        let action = parts[0]
        let start = parts[1].components(separatedBy: ",").map { Int($0)! }
        let end = parts[3].components(separatedBy: ",").map { Int($0)! }
        return Action(act: action, start: (start[0], start[1]), end: (end[0], end[1]))
    }
}

func main(){
    let args = CommandLine.arguments
    if args.count < 2 {
        print("Usage: swift main-part1.swift <input>")
        return
    }

    let filename = args[1]

    let input = try! String(contentsOfFile: filename)


    let lines = input.components(separatedBy: "\n")
    
    var grid = Array(repeating: Array(repeating: false, count: 1000), count: 1000)
    
    for line in lines {
        let action = get_action(line)
        for i in action.start.0...action.end.0 {
            for j in action.start.1...action.end.1 {
                if action.act == "on" {
                    grid[i][j] = true
                } else if action.act == "off" {
                    grid[i][j] = false
                } else {
                    grid[i][j] = !grid[i][j]
                }
            }
        }
    }

    var count = 0
    for i in 0..<1000 {
        for j in 0..<1000 {
            if grid[i][j] {
                count += 1
            }
        }
    }

    print("Part 1 \(count)")
}

main()

File: 2015/typescript/6/main-part-1.ts
import { existsSync, readFileSync, PathLike } from "fs";

type Action = {
  act: string;
  start: number[];
  end: number[];
};

function get_action(line: string): Action {
  let parts: string[] = line.split(" ");
  if (parts[0] == "turn") {
    let act: string = parts[1];
    let start: number[] = parts[2].split(",").map(Number);
    let end: number[] = parts[4].split(",").map(Number);
    return { act: act, start: start, end: end };
  } else {
    let act: string = parts[0];
    let start: number[] = parts[1].split(",").map(Number);
    let end: number[] = parts[3].split(",").map(Number);
    return { act: act, start: start, end: end };
  }
}

function main() {
  let args: string[] = process.argv.slice(1);

  if (args.length < 2) {
    console.log("Usage: node main-part1.js <input.txt>");
    return;
  }

  let filename: string = args[1];
  if (!existsSync(filename)) {
    console.log(`File not found: ${filename}`);
    return;
  }

  let input: string = readFileSync(filename, "utf-8");

  let lines: string[] = input.split("\n");

  let grid: boolean[][] = new Array(1000)
    .fill(false)
    .map(() => new Array(1000).fill(false));

  for (let line of lines) {
    let action: Action = get_action(line);
    let act: string = action.act;
    let start: number[] = action.start;
    let end: number[] = action.end;

    for (let i = start[0]; i <= end[0]; i++) {
      for (let j = start[1]; j <= end[1]; j++) {
        if (act == "on") {
          grid[i][j] = true;
        } else if (act == "off") {
          grid[i][j] = false;
        } else {
          grid[i][j] = !grid[i][j];
        }
      }
    }
  }

  let count: number = 0;
  for (let i = 0; i < 1000; i++) {
    for (let j = 0; j < 1000; j++) {
      if (grid[i][j]) {
        count++;
      }
    }
  }

  console.log("Part 1:", count);
}

main();
